### ISSUE TITLE
instance of typing.Generic class as a typing.ClassVar creates paradox

### ISSUE THREAD

**Issue by raceychan** [microsoft/pyright#6395]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
when use instance of a generic class as classvar, there will be a paradox, where '"ClassVar" type cannot include type variables', but also
'Expected type arguments for generic class' 

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded  by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
import typing as ty

ObjectId = ty.TypeVar("ObjectId", str, int)


def registry_factory() -> dict[ObjectId, "BaseOjbect[ObjectId]"]:
    return dict()


class BaseOjbect(ty.Generic[ObjectId]):
    registry: ty.ClassVar[dict[ObjectId, "BaseOjbect"]] = registry_factory()

    def __init__(self, object_id: ObjectId):
        self._object_id = object_id
        self.registry[object_id] = self

    def get_obj(self, object_id: ObjectId) -> ty.Self:
        return self.registry[object_id]

    def __init_subclass__(cls):
        cls.registry = registry_factory()


class SubOjbect(BaseOjbect[int]):
    def __init__(self, object_id: int):
        super().__init__(object_id)
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

I am running pyright  as a VS Code extension.

**Comment by raceychan** [comment]
a simpler and more common example would be:
```python
T = ty.TypeVar("T")

class Singleton(ty.Generic[T]):
    _instance: ty.ClassVar["Singleton[T]" | None] = None

    @classmethod
    def get_instance(cls) -> ty.Self:
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

```

**Comment by erictraut** [comment]
Pyright is correct here. It is a type violation to use type variables within a class variable. At runtime, all type arguments are erased, so a class represents all subtypes of that class regardless of how it's specialized. That means class variables cannot depend on type arguments. In your example, if you store a `Singleton[int]` instance into `_instance`, that would be a problem if you later referenced `_instance` from a `Singleton[str]` instance.

I'm not sure what you're trying to do here, so I can't offer you advice on a workaround, but you appear to be attempting to do something that is intrinsically unsafe from a type perspective. Pyright is therefore correct in generating an error. Mypy generates the same error in this case.

**Comment by erictraut** [comment]
I'll also note that in your first example, you're using a TypeVar named `ObjectId` in two different contexts — the function `registry_factory` and the class `BaseObject`. Those are two different type variables, and they have different scopes. They have no relation to each other.

Perhaps this approach addresses the problem you're trying to solve?

```python
from typing import ClassVar, Hashable, Self, cast, Any

class BaseObject[T: Hashable]:
    registry: ClassVar[dict[Hashable, "BaseObject[Any]"]]

    def __init__(self, object_id: T):
        self._object_id = object_id
        self.registry[object_id] = self

    def get_obj(self, object_id: T) -> Self | None:
        return cast(Self | None, self.registry[object_id])

    def __init_subclass__(cls):
        cls.registry = dict()

class SubObject(BaseObject[int]):
    def __init__(self, object_id: int):
        super().__init__(object_id)
```

**Comment by raceychan** [comment]
> Pyright is correct here. It is a type violation to use type variables within a class variable. At runtime, all type arguments are erased, so a class represents all subtypes of that class regardless of how it's specialized. That means class variables cannot depend on type arguments. In your example, if you store a `Singleton[int]` instance into `_instance`, that would be a problem if you later referenced `_instance` from a `Singleton[str]` instance.
> 
> I'm not sure what you're trying to do here, so I can't offer you advice on a workaround, but you appear to be attempting to do something that is intrinsically unsafe from a type perspective. Pyright is therefore correct in generating an error. Mypy generates the same error in this case.

Eric, thanks for your quick response and helpful explaination. does this imply that generally we can't use Generic class as a class variable?

**Comment by erictraut** [comment]
> does this imply that generally we can't use Generic class as a class variable

That's correct. You need to specialize a generic class to use it in a class variable, even if that means specializing it with `Any`.

**Comment by raceychan** [comment]
> > does this imply that generally we can't use Generic class as a class variable
> 
> That's correct. You need to specialize a generic class to use it in a class variable, even if that means specializing it with `Any`.

Yeah, Your solution serves my purpose well, I think I just missed the fact that you can't do 

```python
class BaseObject(ty.Generic[T]):
    registry: ClassVar[dict[Hashable, "BaseObject[T]"]]
```
but instead you should do

```python
class BaseObject(ty.Generic[T]):
    registry: ClassVar[dict[Hashable, "BaseObject[Any]"]]
```

However, when I changed T to 
```python
T = TypeVar("T", bound=Any)
```
pyright would still complain. does T here means the same as Any?

**Comment by raceychan** [comment]
> > does this imply that generally we can't use Generic class as a class variable
> 
> That's correct. You need to specialize a generic class to use it in a class variable, even if that means specializing it with `Any`.
I have made a couple attemps trying to replace Any with types like Hashable, str, or int, but they all failed

I was hoping that 
```python
class BaseObject(ty.Generic[T]):
    registry: ClassVar[dict[Hashable, "BaseObject[Hashable]"]]
```
would work, but pyright says

"BaseObject[T@BaseObject]" is incompatible with "BaseObject[Hashable]"
    Type parameter "T@BaseObject" is invariant, but "T@BaseObject" is not the same as "Hashable"

which i don't quite understand

it seems like that Any is the only option we have in situation like this?
